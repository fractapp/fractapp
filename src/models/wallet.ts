import { ImageSourcePropType } from "react-native";

export enum Currency {
    Polkadot = 0,
    Kusama
}

export function getLogo(currency: Currency) {
    let logo: ImageSourcePropType
    switch (currency) {
        case Currency.Polkadot:
            logo = require("assets/img/dot.png")
            break
        case Currency.Kusama:
            logo = require("assets/img/kusama.png")
            break
        default:
            throw ("invalid currency")
    }
    return logo
}
export function getSymbol(currency: Currency) {
    let symbol = ""
    switch (currency) {
        case Currency.Polkadot:
            symbol = "DOT"
            break
        case Currency.Kusama:
            symbol = "KSM"
            break
        default:
            throw ("invalid currency")
    }
    return symbol
}

export class Wallet {
    name: string;
    address: string;
    balance: number;
    currency: Currency;
    usdValue: number;

    constructor(name: string, address: string, currency: Currency, balance: number, price: number) {
        this.name = name;
        this.address = address;
        this.balance = balance;
        this.currency = currency;

        this.usdValue = +(balance * price).toFixed(2)
    }
}
