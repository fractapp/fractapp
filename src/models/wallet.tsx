import { ImageSourcePropType } from "react-native";

export enum Currency {
    Polkadot = 0,
    Kusama
}

export class Wallet {
    name: string;
    address: string;
    balance: number;
    currency: Currency;

    symbol: string;
    usdValue: number;
    logo: ImageSourcePropType;

    constructor(name: string, address: string, currency: Currency, balance: number) {
        this.name = name;
        this.address = address;
        this.balance = balance;
        this.currency = currency;
        
        //TODO get state
        let rate = 10
        this.usdValue = balance * rate
        switch (currency) {
            case Currency.Polkadot:
                this.logo = require("assets/img/dot.png")
                this.symbol = "DOT"
                break
            case Currency.Kusama:
                this.logo = require("assets/img/kusama.png")
                this.symbol = "KSM"
                break
        }

    }
}
