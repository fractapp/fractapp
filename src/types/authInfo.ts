/**
 * @category Models
 */
export type AuthInfo = {
  hasWallet: boolean;
  hasPasscode: boolean;
  hasBiometry: boolean;
};

/**
 * @category Models
 */
export type LoadInfo = {
  isAllStatesLoaded: boolean,
  isSyncShow: boolean;
  isLoadingShow: boolean;
};
