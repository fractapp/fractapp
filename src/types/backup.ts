/**
 * File backup model
 * @category Models
 */
export type FileBackup = {
  seed: string;
  algorithm: string;
};
