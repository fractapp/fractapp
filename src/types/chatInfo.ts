/**
 * @category Models
 */
export type ChatInfo = {
  id: string;
  notificationCount: number;
  lastMsgId: string | null
};
