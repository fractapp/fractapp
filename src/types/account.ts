import {Currency} from './wallet';

export enum Network {
  Polkadot = 0,
  Kusama,
}

/**
 * @category Models
 */
export type Account = {
  name: string;
  address: string;
  pubKey: string;
  currency: Currency;
  network: Network;
  balance: number;
  planks: string;
};
