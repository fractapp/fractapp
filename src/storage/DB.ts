import {Account} from 'models/account';
import AsyncStorage from '@react-native-async-storage/async-storage';
import Keychain from 'react-native-keychain';
import {base64Encode, randomAsU8a} from '@polkadot/util-crypto';
import PasscodeUtil from 'utils/passcode';
import {Keyring} from '@polkadot/keyring';
import {u8aToHex} from '@polkadot/util';
import {Currency, getSymbol} from 'models/wallet';
import {Transaction} from 'models/transaction';
import {ChatInfo} from 'models/chatInfo';
import {AuthInfo} from 'models/authInfo';
import {MyProfile} from 'models/myProfile';
import BN from 'bn.js';
import {UserProfile} from 'models/profile';
import {Use} from 'react-native-svg';
/**
 * @namespace
 * @category storage
 */

namespace DB {
  const secureOption = {
    service: 'com.fractapp',
  };
  const biometryOption = {
    service: 'com.fractapp',
    accessControl: Keychain.ACCESS_CONTROL.BIOMETRY_CURRENT_SET,
  };

  const AsyncStorageKeys = {
    price: (currency: Currency) => `price_${currency}`,
    profile: 'profile',
    authInfo: 'auth_info',
    notificationCount: 'notification_count',
    accounts: 'accounts',
    contacts: 'contacts',
    users: 'users',
    accountInfo: (address: string) => `account_${address}`,
    transactions: (currency: Currency) => `transactions_${getSymbol(currency)}`,
    pendingTransactions: (currency: Currency) =>
      `pending_transactions_${getSymbol(currency)}`,
    chatByAddress: (address: string) => `chat_${address}`,
    chatsInfo: 'chats_info',
  };
  const SecureStorageKeys = {
    firebaseToken: 'firebase_token',
    authJWT: 'auth_jwt',
    seed: 'seed',
    salt: 'salt',
    passcodeHash: 'passcode_hash',
  };
  const PasscodeStorageKeys = {
    passcode: 'passcode',
  };

  export async function setSecureItem(key: string, value: string) {
    const result = await Keychain.setInternetCredentials(
      key,
      key,
      value,
      secureOption,
    );
    if (result == false) {
      throw `invalid set ${key}`;
    }
  }
  export async function getSecureItem(key: string): Promise<string | null> {
    const result = await Keychain.getInternetCredentials(key, secureOption);
    if (result === false) {
      return null;
    }

    return result.password;
  }

  export async function setAuthInfo(auth: AuthInfo) {
    await AsyncStorage.setItem(AsyncStorageKeys.authInfo, JSON.stringify(auth));
  }
  export async function getAuthInfo(): Promise<AuthInfo | null> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.authInfo);
    if (result == null) {
      return null;
    }
    return JSON.parse(result);
  }

  export async function setPrice(currency: Currency, value: number) {
    await AsyncStorage.setItem(AsyncStorageKeys.price(currency), String(value));
  }
  export async function getPrice(currency: Currency): Promise<number> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.price(currency));
    if (result == null) {
      return 0;
    }
    return Number(result);
  }

  export async function setProfile(profile: MyProfile) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.profile,
      JSON.stringify(profile),
    );
  }
  export async function getProfile(): Promise<MyProfile | null> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.profile);
    if (result == null) {
      return null;
    }
    return JSON.parse(result);
  }

  export async function setNotificationCount(count: number) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.notificationCount,
      String(count),
    );
  }
  export async function getNotificationCount(): Promise<number> {
    const result = await AsyncStorage.getItem(
      AsyncStorageKeys.notificationCount,
    );
    return Number(result);
  }

  export async function enablePasscode(passcode: string, isBiometry: boolean) {
    if (isBiometry) {
      const result = await Keychain.setInternetCredentials(
        PasscodeStorageKeys.passcode,
        PasscodeStorageKeys.passcode,
        passcode,
        biometryOption,
      );
      if (result === false) {
        throw 'invalid set passcode';
      }
    }

    const salt = base64Encode(randomAsU8a(32));
    const hash = PasscodeUtil.hash(passcode, salt);
    let authInfo = await getAuthInfo();
    if (authInfo == null) {
      authInfo = new AuthInfo(false, false, false, false);
    }

    try {
      await setSecureItem(SecureStorageKeys.salt, salt);
      await setSecureItem(SecureStorageKeys.passcodeHash, hash);
      authInfo.isPasscode = true;
      authInfo.isBiometry = isBiometry;
    } catch (e) {
      await disablePasscode();
    }

    await setAuthInfo(authInfo);
  }
  export async function disablePasscode() {
    let authInfo = await getAuthInfo();
    if (authInfo == null) {
      authInfo = new AuthInfo(false, false, false, false);
    }
    authInfo.isPasscode = false;
    authInfo.isBiometry = false;

    await setAuthInfo(authInfo);

    await Keychain.resetInternetCredentials(
      SecureStorageKeys.salt,
      secureOption,
    );
    await Keychain.resetInternetCredentials(
      SecureStorageKeys.passcodeHash,
      secureOption,
    );
  }

  export async function getSalt(): Promise<string | null> {
    return await getSecureItem(SecureStorageKeys.salt);
  }
  export async function getPasscodeHash(): Promise<string | null> {
    return await getSecureItem(SecureStorageKeys.passcodeHash);
  }
  export async function getPasscode(): Promise<string> {
    const result = await Keychain.getInternetCredentials(
      PasscodeStorageKeys.passcode,
      biometryOption,
    );
    if (result === false) {
      throw 'value by passcode key not found';
    }

    return result.password;
  }

  export async function createAccounts(seed: string) {
    let polkadotWallet = new Keyring({
      type: 'sr25519',
      ss58Format: 0,
    }).addFromUri(seed);
    let kusamaWallet = new Keyring({type: 'sr25519', ss58Format: 2}).addFromUri(
      seed,
    );
    let accountsInfo = new Array<Account>(
      new Account(
        'Polkadot wallet',
        polkadotWallet.address,
        u8aToHex(polkadotWallet.publicKey),
        Currency.Polkadot,
        0,
        new BN(0).toString(10),
      ),
      new Account(
        'Kusama wallet',
        kusamaWallet.address,
        u8aToHex(kusamaWallet.publicKey),
        Currency.Kusama,
        0,
        new BN(0).toString(10),
      ),
    );
    let accounts = new Array<string>();
    for (let i = 0; i < accountsInfo.length; i++) {
      const account = accountsInfo[i];
      accounts.push(account.address);
      await setAccountInfo(account);
    }
    await setAccounts(accounts);
    await setSecureItem(SecureStorageKeys.seed, seed);
  }
  export async function getSeed(): Promise<string | null> {
    return await getSecureItem(SecureStorageKeys.seed);
  }

  export async function setAccounts(accounts: Array<string>) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.accounts,
      JSON.stringify(accounts),
    );
  }
  export async function getAccounts(): Promise<Array<string> | null> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.accounts);
    if (result == null) {
      return null;
    }
    return JSON.parse(result);
  }

  export async function setChatsInfo(chats: Map<string, ChatInfo>) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.chatsInfo,
      JSON.stringify([...chats]),
    );
  }
  export async function getChatsInfo(): Promise<Map<string, ChatInfo>> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.chatsInfo);

    if (result == null) {
      return new Map<string, ChatInfo>();
    }
    return new Map<string, ChatInfo>(JSON.parse(result));
  }

  export async function setChat(address: string, chats: Map<string, Currency>) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.chatByAddress(address),
      JSON.stringify([...chats]),
    );
  }
  export async function getChat(
    address: string,
  ): Promise<Map<string, boolean>> {
    const result = await AsyncStorage.getItem(
      AsyncStorageKeys.chatByAddress(address),
    );

    if (result == null) {
      return new Map<string, boolean>();
    }
    return new Map<string, boolean>(JSON.parse(result));
  }

  export async function setTx(currency: Currency, tx: Transaction) {
    const txs = await getTxs(currency);

    txs.set(tx.id, tx);
    await AsyncStorage.setItem(
      AsyncStorageKeys.transactions(currency),
      JSON.stringify([...txs]),
    );
  }
  export async function getTxs(
    currency: Currency,
  ): Promise<Map<string, Transaction>> {
    const result = await AsyncStorage.getItem(
      AsyncStorageKeys.transactions(currency),
    );

    if (result == null) {
      return new Map<string, Transaction>();
    }
    return new Map<string, Transaction>(JSON.parse(result));
  }

  export async function setContacts(contacts: Array<UserProfile>) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.contacts,
      JSON.stringify(contacts),
    );
  }

  export async function getContacts(): Promise<Array<UserProfile>> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.contacts);

    if (result == null) {
      return [];
    }

    return JSON.parse(result);
  }

  export async function setUsers(users: Map<string, UserProfile>) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.users,
      JSON.stringify([...users]),
    );
  }

  export async function getUsers(): Promise<Map<string, UserProfile>> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.users);

    if (result == null) {
      return new Map<string, UserProfile>();
    }
    return new Map<string, UserProfile>(JSON.parse(result));
  }

  export async function setPendingTxs(currency: Currency, txs: Array<string>) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.pendingTransactions(currency),
      JSON.stringify(txs),
    );
  }

  export async function getPendingTxs(
    currency: Currency,
  ): Promise<Array<string>> {
    const result = await AsyncStorage.getItem(
      AsyncStorageKeys.pendingTransactions(currency),
    );

    if (result == null) {
      return [];
    }

    return JSON.parse(result);
  }

  export async function setAccountInfo(account: Account) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.accountInfo(account.address),
      JSON.stringify(account),
    );
  }
  export async function getAccountInfo(
    address: string,
  ): Promise<Account | null> {
    const result = await AsyncStorage.getItem(
      AsyncStorageKeys.accountInfo(address),
    );
    if (result == null) {
      return null;
    }
    return JSON.parse(result);
  }

  export async function setFirebaseToken(token: string) {
    await setSecureItem(SecureStorageKeys.firebaseToken, token);
  }
  export async function getFirebaseToken(): Promise<string | null> {
    return await getSecureItem(SecureStorageKeys.firebaseToken);
  }

  export async function setJWT(token: string) {
    await setSecureItem(SecureStorageKeys.authJWT, token);
  }
  export async function getJWT(): Promise<string | null> {
    return await getSecureItem(SecureStorageKeys.authJWT);
  }
}

export default DB;
