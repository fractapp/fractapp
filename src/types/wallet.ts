import { ImageSourcePropType } from 'react-native';
import { AccountType, Network } from 'types/account';
import { Transaction, TxAction } from 'types/transaction';

/**
 * @category Models
 */
export enum Currency {
  DOT = 0,
  KSM,
}


/**
 * get network by currency
 * @category Models
 */

export function getNetwork(currency: Currency): Network {
  switch (currency) {
    case Currency.DOT:
      return Network.Polkadot;
    case Currency.KSM:
      return Network.Kusama;
    default:
      return Network.Polkadot;
  }
}


/**
 * get logo for wallet
 * @category Models
 */

export function getFullCurrencyName(currency: Currency) {
  let name = '';
  switch (currency) {
    case Currency.DOT:
      name = 'Polkadot';
      break;
    case Currency.KSM:
      name = 'Kusama';
      break;
    default:
      throw new Error('invalid currency');
  }
  return name;
}

export function getColor(currency: Currency) {
  let color = '';
  switch (currency) {
    case Currency.DOT:
      color = '#E6007A';
      break;
    case Currency.KSM:
      color = '#434C74';
      break;
    default:
      throw new Error('invalid currency');
  }
  return color;
}

export function getLogo(currency: Currency) {
  let logo: ImageSourcePropType;
  switch (currency) {
    case Currency.DOT:
      logo = require('assets/img/dot.png');
      break;
    case Currency.KSM:
      logo = require('assets/img/kusama.png');
      break;
    default:
      throw new Error('invalid currency');
  }
  return logo;
}

/**
 * check logo without border
 * @category Models
 */
export function withoutBorder(currency: Currency): boolean {
  return currency === Currency.DOT || currency === Currency.KSM;
}

/**
 * get symbol for wallet
 * @category Models
 */
export function getSymbol(currency: Currency) {
  let symbol = '';
  switch (currency) {
    case Currency.DOT:
      symbol = 'DOT';
      break;
    case Currency.KSM:
      symbol = 'KSM';
      break;
    default:
      throw new Error('invalid currency');
  }
  return symbol;
}

/**
 * string to currency
 * @category Models
 */
export function toCurrency(currency: string): Currency {
  switch (currency) {
    case 'DOT':
      return Currency.DOT;
    case 'KSM':
      return Currency.KSM;
    default:
      throw new Error('invalid currency');
  }
}


/**
 * currency to string
 * @category Models
 */
export function fromCurrency(currency: Currency): string {
  switch (currency) {
    case Currency.DOT:
      return 'DOT';
    case Currency.KSM:
      return 'KSM';
    default:
      throw new Error('invalid currency');
  }
}

/**
 * filter Tx by wallet
 * @category Models
 */
export function filterTxsByAccountType(txs: Array<Transaction>, action: AccountType): Array<Transaction>  {
  switch (action) {
    case AccountType.Main:
      return txs.filter((tx) => tx.action === TxAction.Transfer || tx.action === TxAction.Staking || tx.action === TxAction.StakingWithdrawn);
    case AccountType.Staking:
      return txs.filter((tx) => tx.action === TxAction.Staking || tx.action === TxAction.StakingWithdrawn || tx.action === TxAction.StakingReward);
    default:
      throw new Error('invalid action type');
  }
}
