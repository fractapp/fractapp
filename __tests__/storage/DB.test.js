import Keychain from 'react-native-keychain';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {Currency} from 'types/wallet';
import DB from 'storage/DB';
import AccountsStore from 'storage/Accounts';
import ChatsStore from 'storage/Chats';

jest.mock('react-native-keychain', () => ({
  setInternetCredentials: jest.fn(),
  getInternetCredentials: jest.fn(),
  ACCESS_CONTROL: {
    BIOMETRY_CURRENT_SET: 0,
  },
}));
jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(),
  setItem: jest.fn(),
}));
jest.mock('@polkadot/util-crypto', () => ({
  setAccountInfo: jest.fn(),
}));
jest.mock('@polkadot/keyring', () => {});
jest.mock('@polkadot/util', () => ({
  u8aToHex: jest.fn(),
}));
jest.mock('utils/passcode', () => ({
  hash: jest.fn(),
}));
jest.mock('@polkadot/util-crypto', () => ({
  base64Encode: jest.fn(),
  randomAsU8a: jest.fn(),
}));
jest.mock('react-native-i18n', () => ({
  t: (value) => value,
}));


it('Test secureOption', async () => {
  expect(DB.secureOption).toMatchSnapshot();
});
it('Test AsyncStorageKeys', async () => {
  expect(DB.AsyncStorageKeys).toMatchSnapshot();
});
it('Test SecureStorageKeys', async () => {
  expect(DB.SecureStorageKeys).toMatchSnapshot();
});
it('Test PasscodeStorageKeys', async () => {
  expect(DB.PasscodeStorageKeys).toMatchSnapshot();
});

it('Test setSecureItem positive', async () => {
  const key = 'key';
  const value = 'value';
  await DB.setSecureItem(key, value);
  expect(Keychain.setInternetCredentials).toBeCalledWith(
    key,
    key,
    value,
    DB.secureOption,
  );
});
it('Test setSecureItem negative', async () => {
  const key = 'key';
  const value = 'value';
  Keychain.setInternetCredentials.mockReturnValueOnce(false);
  await expect(DB.setSecureItem(key, value)).rejects.toThrow(
    `invalid set ${key}`,
  );

  expect(Keychain.setInternetCredentials).toBeCalledWith(
    key,
    key,
    value,
    DB.secureOption,
  );
});
it('Test getSecureItem positive', async () => {
  const key = 'key';
  const value = 'value';
  Keychain.getInternetCredentials.mockReturnValueOnce({
    password: value,
  });
  const item = await DB.getSecureItem(key);
  expect(item).toEqual(value);
  expect(Keychain.getInternetCredentials).toBeCalledWith(key, DB.secureOption);
});
it('Test getSecureItem negative', async () => {
  const key = 'key';
  Keychain.getInternetCredentials.mockReturnValueOnce(false);
  const item = await DB.getSecureItem(key);
  expect(item).toEqual(null);
  expect(Keychain.getInternetCredentials).toBeCalledWith(key, DB.secureOption);
});

it('Test setAuthInfo', async () => {
  const auth = {
    isSynced: false,
    isAuthed: true,
    isPasscode: false,
    isBiometry: true,
  };
  await DB.setAuthInfo(auth);
  expect(AsyncStorage.setItem).toBeCalledWith(
    DB.AsyncStorageKeys.authInfo,
    JSON.stringify(auth),
  );
});
it('Test getAuthInfo negative', async () => {
  AsyncStorage.getItem.mockReturnValueOnce(null);

  const value = await DB.getAuthInfo();
  expect(value).toEqual(null);
});
it('Test getAuthInfo positive', async () => {
  const auth = {
    isFirstSync: true,
    hasWallet: true,
    hasPasscode: true,
    hasBiometry: true,
  };

  AsyncStorage.getItem.mockReturnValueOnce(JSON.stringify(auth));
  const value = await DB.getAuthInfo();
  expect(value).toEqual(auth);
});


it('Test setProfile', async () => {
  const profile = {
    id: 'id',
    name: 'name',
    username: 'username',
    phoneNumber: 'phone',
    email: 'email',
    avatarExt: 'png',
    lastUpdate: 123,
  };
  await DB.setProfile(profile);
  expect(AsyncStorage.setItem).toBeCalledWith(
    DB.AsyncStorageKeys.profile,
    JSON.stringify(profile),
  );
});
it('Test getProfile negative', async () => {
  AsyncStorage.getItem.mockReturnValueOnce(null);

  const value = await DB.getProfile();
  expect(value).toEqual(null);
});
it('Test getProfile positive', async () => {
  const profile = {
    id: 'id',
    name: 'name',
    username: 'username',
    phoneNumber: 'phone',
    email: 'email',
    avatarExt: 'png',
    lastUpdate: 123,
  };

  AsyncStorage.getItem.mockReturnValueOnce(JSON.stringify(profile));
  const p = await DB.getProfile();
  expect(p).toEqual(profile);
  expect(AsyncStorage.getItem).toBeCalledWith(DB.AsyncStorageKeys.profile);
});

it('Test getSalt', async () => {
  const v = 'value';
  Keychain.getInternetCredentials.mockReturnValueOnce({
    password: v,
  });
  const salt = await DB.getSalt();
  expect(salt).toEqual(v);
  expect(Keychain.getInternetCredentials).toBeCalledWith(
    DB.SecureStorageKeys.salt,
    DB.secureOption,
  );
});
it('Test getPasscodeHash', async () => {
  const v = 'value';
  Keychain.getInternetCredentials.mockReturnValueOnce({
    password: v,
  });
  const hash = await DB.getPasscodeHash();
  expect(hash).toEqual(v);
  expect(Keychain.getInternetCredentials).toBeCalledWith(
    DB.SecureStorageKeys.passcodeHash,
    DB.secureOption,
  );
});
it('Test getSeed', async () => {
  const v = 'value';
  Keychain.getInternetCredentials.mockReturnValueOnce({
    password: v,
  });
  const seed = await DB.getSeed();
  expect(seed).toEqual(v);
  expect(Keychain.getInternetCredentials).toBeCalledWith(
    DB.SecureStorageKeys.seed,
    DB.secureOption,
  );
});

it('Test setAccounts', async () => {
  const state = AccountsStore.initialState();
  await DB.setAccountStore(state);
  expect(AsyncStorage.setItem).toBeCalledWith(
    DB.AsyncStorageKeys.accountsStore,
    JSON.stringify(state),
  );
});
it('Test getAccounts positive', async () => {
  const state = AccountsStore.initialState();
  AsyncStorage.getItem.mockReturnValueOnce(JSON.stringify(state));

  const a = await DB.getAccountStore();
  expect(a).toEqual(state);
  expect(AsyncStorage.getItem).toBeCalledWith(DB.AsyncStorageKeys.accountsStore);
});
it('Test getAccounts negative', async () => {
  AsyncStorage.getItem.mockReturnValueOnce(null);

  const value = await DB.getAccountStore();
  expect(value).toEqual(null);
});

it('Test getChatsState positive', async () => {
  const state = ChatsStore.initialState();

  AsyncStorage.getItem.mockReturnValueOnce(JSON.stringify(state));

  const c = await DB.getChatsState();
  expect(c).toEqual(state);
  expect(AsyncStorage.getItem).toBeCalledWith(DB.AsyncStorageKeys.chatsStorage);
});

it('Test setChatsState', async () => {
  const state = ChatsStore.initialState();

  await DB.setChatsState(state);
  expect(AsyncStorage.setItem).toBeCalledWith(
    DB.AsyncStorageKeys.chatsStorage,
    JSON.stringify(state),
  );
});

it('Test setContacts', async () => {
  const contacts = [ 'a', 'b' ];

  await DB.setContacts(contacts);
  expect(AsyncStorage.setItem).toBeCalledWith(
    DB.AsyncStorageKeys.contacts,
    JSON.stringify(contacts),
  );
});
it('Test getContacts positive', async () => {
  const contacts = [ 'a', 'b' ];

  AsyncStorage.getItem.mockReturnValueOnce(JSON.stringify(contacts));

  const c = await DB.getContacts();
  expect(c).toEqual(contacts);
  expect(AsyncStorage.getItem).toBeCalledWith(DB.AsyncStorageKeys.contacts);
});
it('Test getContacts negative', async () => {
  AsyncStorage.getItem.mockReturnValueOnce(null);

  const value = await DB.getContacts();
  expect(value).toEqual([]);
});

it('Test setUsers', async () => {
 const users = {
   'userId': {
     isAddressOnly: true,
     title: 'title',
     value: {
       address: 'address',
       currency: Currency.DOT,
     },
   },
 };

  await DB.setUsers(users);
  expect(AsyncStorage.setItem).toBeCalledWith(
    DB.AsyncStorageKeys.users,
    JSON.stringify(users),
  );
});
it('Test getUsers positive', async () => {
  const users = {
    'userId': {
      isAddressOnly: true,
      title: 'title',
      value: {
        address: 'address',
        currency: Currency.DOT,
      },
    },
  };

  AsyncStorage.getItem.mockReturnValueOnce(JSON.stringify(users));

  const u = await DB.getUsers();
  expect(u).toEqual(users);
  expect(AsyncStorage.getItem).toBeCalledWith(DB.AsyncStorageKeys.users);
});
it('Test getUsers negative', async () => {
  AsyncStorage.getItem.mockReturnValueOnce(null);

  const value = await DB.getUsers();
  expect(value).toEqual({});
});

it('Test setFirebaseToken', async () => {
  const v = 'value';
  await DB.setFirebaseToken(v);
  expect(Keychain.setInternetCredentials).toBeCalledWith(
    DB.SecureStorageKeys.firebaseToken,
    DB.SecureStorageKeys.firebaseToken,
    v,
    DB.secureOption,
  );
});
it('Test getFirebaseToken', async () => {
  const v = 'value';
  Keychain.getInternetCredentials.mockReturnValueOnce({
    password: v,
  });
  const token = await DB.getFirebaseToken();
  expect(token).toEqual(v);
  expect(Keychain.getInternetCredentials).toBeCalledWith(
    DB.SecureStorageKeys.firebaseToken,
    DB.secureOption,
  );
});

it('Test setJWT', async () => {
  const v = 'value';
  await DB.setJWT(v);
  expect(Keychain.setInternetCredentials).toBeCalledWith(
    DB.SecureStorageKeys.authJWT,
    DB.SecureStorageKeys.authJWT,
    v,
    DB.secureOption,
  );
});
it('Test getJWT', async () => {
  const v = 'value';
  Keychain.getInternetCredentials.mockReturnValueOnce({
    password: v,
  });
  const token = await DB.getJWT();
  expect(token).toEqual(v);
  expect(Keychain.getInternetCredentials).toBeCalledWith(
    DB.SecureStorageKeys.authJWT,
    DB.secureOption,
  );
});
