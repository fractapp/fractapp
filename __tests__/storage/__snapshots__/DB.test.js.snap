// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test AsyncStorageKeys 1`] = `
Object {
  "accountsStore": "accounts_store",
  "authInfo": "auth_info",
  "chatsStorage": "chatsStorage",
  "contacts": "contacts",
  "lastInitMsgTimeout": "lastInitMsgTimeout",
  "profile": "profile",
  "serverInfo": "server_info",
  "users": "users",
  "version": "version",
}
`;

exports[`Test PasscodeStorageKeys 1`] = `
Object {
  "passcode": "passcode",
}
`;

exports[`Test SecureStorageKeys 1`] = `
Object {
  "authJWT": "auth_jwt",
  "firebaseToken": "firebase_token",
  "passcodeHash": "passcode_hash",
  "salt": "salt",
  "seed": "seed",
}
`;

exports[`Test secureOption 1`] = `
Object {
  "service": "com.fractapp",
}
`;
